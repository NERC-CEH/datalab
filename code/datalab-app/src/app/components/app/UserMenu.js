import React from 'react';
import PropTypes from 'prop-types';
import Avatar from 'material-ui/Avatar';
import Typography from 'material-ui/Typography';
import Button from 'material-ui/Button';
import { withStyles } from 'material-ui/styles';
import auth from '../../auth/auth';

const styles = theme => ({
  container: {
    display: 'flex',
    flexDirection: 'column',
    backgroundColor: theme.palette.backgroundColor,
  },
  identity: {
    margin: 10,
    display: 'flex',
    flexDirection: 'row',
  },
  userImage: {
    height: 96,
    width: 96,
  },
  userInfo: {
    marginLeft: 10,
  },
});

const UserMenu = ({ classes, identity, closePopover }) => (
  <div className={classes.container}>
    <div className={classes.identity}>
      <Avatar className={classes.userImage} src={identity.picture} />
      <div className={classes.userInfo}>
        <Typography type="body1">{identity.nickname}</Typography>
        <Typography type="body2">{identity.name}</Typography>
      </div>
    </div>
    <Button
      raised
      color="primary"
      onClick={closePopover(auth.logout)}
    >
      Logout
    </Button>
  </div>
);

UserMenu.propTypes = {
  identity: PropTypes.shape({
    name: PropTypes.string.isRequired,
    nickname: PropTypes.string.isRequired,
    picture: PropTypes.string.isRequired,
  }),
  closePopover: PropTypes.func.isRequired,
};

export default withStyles(styles)(UserMenu);
