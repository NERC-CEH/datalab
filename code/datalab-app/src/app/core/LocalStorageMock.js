class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key];
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export default LocalStorageMock;
