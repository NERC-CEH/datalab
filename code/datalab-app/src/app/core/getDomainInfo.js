function extendSubdomain(extension, localhostPort, location) {
  const domainInfo = getDomainInfo(location);

  if (domainInfo === 'localhost') {
    return `http://localhost:${localhostPort}`;
  }

  return `${domainInfo.protocol}//${domainInfo.subdomain}-${extension}.${domainInfo.domain}`;
}

function replaceSubdomain(subdomain, altHref, location) {
  const domainInfo = getDomainInfo(location);

  if (domainInfo === 'localhost') {
    return altHref || 'localhost';
  }

  return `${domainInfo.protocol}//${subdomain}.${domainInfo.domain}`;
}

function getDomainInfo(location) {
  const currentLocation = location || window.location;

  if (currentLocation.hostname === 'localhost') {
    return 'localhost';
  }

  const { protocol } = currentLocation;
  const { hostname } = currentLocation;
  const subdomain = hostname.split('.')[0];
  const domain = hostname.substring(subdomain.length + 1, hostname.length);

  return { protocol, subdomain, domain };
}

export { getDomainInfo, replaceSubdomain, extendSubdomain };
