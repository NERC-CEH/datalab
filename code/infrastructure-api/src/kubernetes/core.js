import logger from 'winston';
import { has, get } from 'lodash';

export const handleCreateError = (type, name) => (error) => {
  if (has(error, 'response.data.message')) {
    throw createError(`Unable to create kubernetes ${type} '${name}' - ${error.response.data.message}`);
  }
  throw createError(`unable to create kubernetes ${type} '${name}' - ${error}`);
};

export const handleDeleteError = (type, name) => (error) => {
  if (has(error, 'response.status') && get(error, 'response.status') === 404) {
    logger.warn(`Kubernetes API: Could not find ${type}: ${name} to delete it`);
    return Promise.resolve();
  }

  logger.error(`Error deleting ${type}: ${name}`);
  throw createError(error.message);
};

function createError(message) {
  return new Error(`Kubernetes API: ${message}`);
}
