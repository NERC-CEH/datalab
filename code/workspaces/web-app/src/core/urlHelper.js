export default function getUrlNameStartEndText(projectKey, windowLocation) {
  const separator = '.';
  const restHostname = windowLocation.hostname.split(separator).slice(1);
  const startText = `${windowLocation.protocol}//${projectKey}-`;

  let endText = `${separator}${restHostname.join(separator)}`;

  if (windowLocation.hostname === 'localhost') {
    endText = '.datalabs.localhost';
  }

  return { startText, endText };
}
