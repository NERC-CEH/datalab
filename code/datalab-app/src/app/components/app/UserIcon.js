import React, { Component } from 'react';
import { findDOMNode } from 'react-dom';
import PropTypes from 'prop-types';
import Avatar from '@material-ui/core/Avatar';
import Popover from '@material-ui/core/Popover';
import UserMenu from './UserMenu';

class UserIcon extends Component {
  constructor(props, context) {
    super(props, context);
    this.closeOnClick = this.closeOnClick.bind(this);
    this.togglePopup = this.togglePopup.bind(this);
    this.state = {
      open: false,
      anchorEl: null,
    };
  }

  togglePopup = () => {
    this.setState({
      open: !this.state.open,
      anchorEl: this.state.open ? null : findDOMNode(this.userImage),
    });
  };

  closeOnClick = next => () => {
    this.setState({ open: false });
    next();
  };

  render() {
    return (
      <div>
        <Avatar
          ref={(node) => { this.userImage = node; }}
          onClick={this.togglePopup}
          src={this.props.identity.picture}
        />
        <Popover
          open={this.state.open}
          onClose={this.togglePopup}
          anchorEl={this.state.anchorEl}
          anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
        >
          <UserMenu identity={this.props.identity} closePopover={this.closeOnClick} />
        </Popover>
      </div>
    );
  }
}

UserIcon.propTypes = {
  identity: PropTypes.shape({
    picture: PropTypes.string.isRequired,
  }),
};

export default UserIcon;
