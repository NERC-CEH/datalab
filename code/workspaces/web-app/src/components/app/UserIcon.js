import React, { Component } from 'react';
import { findDOMNode } from 'react-dom';
import PropTypes from 'prop-types';
import Avatar from '@material-ui/core/Avatar';
import Popover from '@material-ui/core/Popover';
import { withStyles } from '@material-ui/core/styles';
import UserMenu from './UserMenu';

const styles = theme => ({
  avatar: {
    width: theme.shape.topBarContentHeight,
    height: theme.shape.topBarContentHeight,
    cursor: 'pointer',
  },
});

class UserIcon extends Component {
  constructor(props, context) {
    super(props, context);
    this.closeOnClick = this.closeOnClick.bind(this);
    this.togglePopup = this.togglePopup.bind(this);
    this.state = {
      open: false,
      anchorEl: null,
    };
  }

  togglePopup = () => {
    this.setState({
      open: !this.state.open,
      anchorEl: this.state.open ? null : findDOMNode(this.userImage),
    });
  };

  closeOnClick = next => () => {
    this.setState({ open: false });
    next();
  };

  render() {
    return (
      <div>
        <Avatar
          className={this.props.classes.avatar}
          ref={(node) => { this.userImage = node; }}
          onClick={this.togglePopup}
          src={this.props.identity.picture}
        />
        <Popover
          open={this.state.open}
          onClose={this.togglePopup}
          anchorEl={this.state.anchorEl}
          anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
        >
          <UserMenu identity={this.props.identity} closePopover={this.closeOnClick} />
        </Popover>
      </div>
    );
  }
}

UserIcon.propTypes = {
  identity: PropTypes.shape({
    picture: PropTypes.string.isRequired,
  }),
};

export default withStyles(styles)(UserIcon);
