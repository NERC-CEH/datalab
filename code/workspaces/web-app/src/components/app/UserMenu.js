import React from 'react';
import PropTypes from 'prop-types';
import Avatar from '@material-ui/core/Avatar';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import { withStyles } from '@material-ui/core/styles';
import { getAuth } from '../../config/auth';

const styles = theme => ({
  container: {
    display: 'flex',
    flexDirection: 'column',
    backgroundColor: theme.palette.backgroundColor,
  },
  identity: {
    margin: 10,
    display: 'flex',
    flexDirection: 'row',
  },
  userImage: {
    height: 96,
    width: 96,
  },
  userInfo: {
    marginLeft: 10,
  },
});

const UserMenu = ({ classes, identity, closePopover }) => (
  <div className={classes.container}>
    <div className={classes.identity}>
      <Avatar className={classes.userImage} src={identity.picture} />
      <div className={classes.userInfo}>
        <Typography variant="body1">{identity.nickname}</Typography>
        <Typography variant="body1">{identity.name}</Typography>
      </div>
    </div>
    <Button
      color="primary"
      onClick={closePopover(getAuth().logout)}
    >
      Logout
    </Button>
  </div>
);

UserMenu.propTypes = {
  identity: PropTypes.shape({
    name: PropTypes.string.isRequired,
    nickname: PropTypes.string.isRequired,
    picture: PropTypes.string.isRequired,
  }),
  closePopover: PropTypes.func.isRequired,
};

export default withStyles(styles)(UserMenu);
