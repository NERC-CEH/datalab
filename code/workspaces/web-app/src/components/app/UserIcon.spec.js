import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import Avatar from '@material-ui/core/Avatar';
import Popover from '@material-ui/core/Popover';
import UserMenu from './UserMenu';
import UserIcon from './UserIcon';

const expectedProps = {
  identity: {
    picture: 'expectedPicture',
  },
};

describe('UserIcon', () => {
  let shallow;

  beforeEach(() => {
    shallow = createShallow({ dive: true });
  });

  function shallowRender(props) {
    return shallow(<UserIcon {...props} />);
  }

  it('renders correct snapshot', () => {
    expect(shallowRender(expectedProps)).toMatchSnapshot();
  });

  it('onClick function changes popover open prop', () => {
    const output = shallowRender(expectedProps);
    const onClickFunction = output.find(Avatar).prop('onClick');

    // Due to changes to state, the same node need to be used for the expect checks.
    expect(output.find(Popover).prop('open')).toBe(false);
    onClickFunction();
    expect(output.find(Popover).prop('open')).toBe(true);
  });

  it('closeOnClick function changes popover open prop', () => {
    const output = shallowRender(expectedProps);
    const onClickFunction = output.find(Avatar).prop('onClick');
    const nextMock = jest.fn();
    const closeFunct = output.find(UserMenu).prop('closePopover')(nextMock);

    // Due to changes to state, the same node need to be used for the expect checks.
    onClickFunction(); // should open popover
    expect(output.find(Popover).prop('open')).toBe(true);
    expect(nextMock).not.toHaveBeenCalled();
    closeFunct(); // should close popover
    expect(output.find(Popover).prop('open')).toBe(false);
    expect(nextMock).toHaveBeenCalledTimes(1);
  });
});
