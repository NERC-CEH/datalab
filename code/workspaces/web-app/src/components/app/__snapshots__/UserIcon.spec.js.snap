// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserIcon renders correct snapshot 1`] = `
<div>
  <WithStyles(ForwardRef(Avatar))
    className="UserIcon-avatar-1"
    onClick={[Function]}
    src="expectedPicture"
  />
  <WithStyles(ForwardRef(Popover))
    anchorEl={null}
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "bottom",
      }
    }
    onClose={[Function]}
    open={false}
  >
    <WithStyles(UserMenu)
      closePopover={[Function]}
      identity={
        Object {
          "picture": "expectedPicture",
        }
      }
    />
  </WithStyles(ForwardRef(Popover))>
</div>
`;
